test/shared/src/main/scala/zio/test/BoolAlgebra.scala

```diff
@@ -0,0 +1,327 @@
+        equal(other) ||
+          commutative(other) ||
+          symmetric(associative)(self, other) ||
+          symmetric(distributive)(self, other) ||
+          doubleNegative(self, other) ||
+          deMorgansLaws(other)
+      case _ => false
+    }
+    private def equal(that: BoolAlgebra[_]): Boolean = (self, that) match {
+      case (o1: Or[_], o2: Or[_]) => o1.left == o2.left && o1.right == o2.right
+      case _                      => false
+    }
+    private def associative(left: BoolAlgebra[_], right: BoolAlgebra[_]): Boolean =
+      (left, right) match {
+        case (Or(Or(a1, b1), c1), Or(a2, Or(b2, c2))) =>
+          a1 == a2 && b1 == b2 && c1 == c2
+        case _ =>
+          false
+      }
+    private def commutative(that: BoolAlgebra[_]): Boolean = (self, that) match {
+      case (Or(al, bl), Or(ar, br)) => al == br && bl == ar
+      case _                        => false
+    }
+    private def distributive(left: BoolAlgebra[_], right: BoolAlgebra[_]): Boolean =
+      (left, right) match {
+        case (Or(a1, And(b1, c1)), And(Or(a2, b2), Or(a3, c2))) =>
+          a1 == a2 && a1 == a3 && b1 == b2 && c1 == c2
+        case _ =>
+          false
+      }
+    private def deMorgansLaws(that: BoolAlgebra[_]): Boolean = (self, that) match {
+      case (Or(Not(a), Not(b)), Not(And(c, d))) => a == c && b == d
+      case _                                    => false
+    }
+  }
+
+  final case class Not[+A](result: BoolAlgebra[A]) extends BoolAlgebra[A] { self =>
+    override final def equals(that: Any): Boolean = that match {
+      case other: BoolAlgebra[_] =>
+        equal(other) ||
+          doubleNegative(other, self) ||
+          deMorgansLaws(other)
+      case _ =>
+        false
+    }
+    private def equal(that: BoolAlgebra[_]): Boolean = (self, that) match {
+      case (n1: Not[_], n2: Not[_]) => n1.result == n2.result
+      case _                        => false
+    }
+    private def deMorgansLaws(that: BoolAlgebra[_]): Boolean =
+      (self, that) match {
+        case (Not(Or(a, b)), And(Not(c), Not(d))) => a == c && b == d
+        case (Not(And(a, b)), Or(Not(c), Not(d))) => a == c && b == d
+        case _                                    => false
+      }
+  }
+
+  /**
+   * Returns a result that is the logical conjunction of all of the results in
+   * the specified collection.
+   */
+  final def all[A](as: Iterable[BoolAlgebra[A]]): Option[BoolAlgebra[A]] =
+    if (as.isEmpty) None else Some(as.reduce(_ && _))
+
+  /**
+   * Constructs a result that is the logical conjunction of two results.
+   */
+  final def and[A](left: BoolAlgebra[A], right: BoolAlgebra[A]): BoolAlgebra[A] =
+    And(left, right)
+
+  /**
+   * Returns a result that is the logical disjunction of all of the results in
+   * the specified collection.
+   */
+  final def any[A](as: Iterable[BoolAlgebra[A]]): Option[BoolAlgebra[A]] =
+    if (as.isEmpty) None else Some(as.reduce(_ || _))
```

Comment:
same here, there is nothing to prove so the empty case is failure.

Reply:
ha, same thing as before, you need an `A` for the failure. ok scratch that then.

Reply:
Yeah, exactly.

Reply:
if I'm correct, it's not strictly a bool algebra though, as there should be one and only one "`false`" element.

Reply:
Yeah, I'm definitely not an expert on this. I think in boolean algebra as a branch of mathematics there is only one `true` and one `false` element. I was more trying to go for the idea that it is a reification, an "algebra", of boolean operations. I'm not committed to the name though not sure about a better one. Any ideas?

Reply:
oooh yeah, that makes sense. I don't have a better proposition either way 😆